sklearn/cluster/_hierarchical.pyx

```diff
@@ -449,3 +452,111 @@ def single_linkage_label(L):
+    tree for computation of single linkage clustering. This
+    represents the MST-LINKAGE-CORE algorithm (Figure 6) from
+    *Modern hierarchical, agglomerative clustering algorithms*
+    by Daniel Mullner (https://arxiv.org/abs/1109.2378).
+
+    In contrast to the scipy implementation is never computes
+    a full distance matrix, generating distances only as they
+    are needed and releasing them when no longer needed.
+
+    Parameters
+    ----------
+    raw_data: array of shape (n_samples, n_features)
+        The array of feature data to be clustered.
+
+    dist_metric: DistanceMetric
+        A distance metric function (from sklearn.neighbors.dist_metrics)
+        used to compute distances.
+
+    Returns
+    -------
+    mst_core_data: array of shape (n_samples, 3)
+        An array providing information from which one
+        can either compute an MST, or the linkage hierarchy
+        very efficiently. See https://arxiv.org/abs/1109.2378
+        algorithm MST-LINKAGE-CORE for more details.
+    """
+
+    cdef DTYPE_t[:] current_distances
+    cdef DTYPE_t * raw_data_ptr
+    cdef np.int8_t[:] in_tree
+    cdef DTYPE_t[:, ::1] raw_data_view
+    cdef DTYPE_t[:, ::1] result
+
+    cdef np.ndarray label_filter
+
+    cdef ITYPE_t current_node
+    cdef ITYPE_t new_node
+    cdef ITYPE_t i
+    cdef ITYPE_t j
+    cdef ITYPE_t dim
+    cdef ITYPE_t num_features
+
+    cdef DTYPE_t right_value
+    cdef DTYPE_t left_value
+    cdef DTYPE_t new_distance
+
+    dim = raw_data.shape[0]
+    num_features = raw_data.shape[1]
+
+    raw_data_view = (<DTYPE_t[:raw_data.shape[0], :raw_data.shape[1]:1]> (
+        <DTYPE_t *> raw_data.data))
+    raw_data_ptr = (<DTYPE_t *> &raw_data_view[0, 0])
+
+    result = np.zeros((dim - 1, 3))
+    in_tree = np.zeros(dim, dtype=np.int8)
+    current_node = 0
+    current_distances = np.infty * np.ones(dim)
+
+    for i in range(1, dim):
+
+        in_tree[current_node] = 1
+
+        new_distance = DBL_MAX
+        new_node = 0
+
+        for j in range(dim):
+            if in_tree[j]:
+                continue
+
+            right_value = current_distances[j]
+            left_value = dist_metric.dist(&raw_data_ptr[num_features *
+                                                        current_node],
+                                          &raw_data_ptr[num_features * j],
+                                          num_features)
+
+            if left_value > right_value:
+                if right_value < new_distance:
+                    new_distance = right_value
+                    new_node = j
+                continue
+
+            if left_value < right_value:
+                current_distances[j] = left_value
+                if left_value < new_distance:
+                    new_distance = left_value
+                    new_node = j
+            else:
+                if right_value < new_distance:
+                    new_distance = right_value
+                    new_node = j
```

Comment:
as far as I understand the logic here, you're doing all this to keep track of new_node = argmin_j(current_distances) right?

I might be missing something but there is probably a way to simplify this a bit?

Reply:
It's been a while since I wrote this, but my recollection was that there were simpler approaches, but this was (non-trivially) faster (for, overall, not too much extra complexity). I do know I spent quite a while fiddling with exactly how this worked before arriving at this as the best option.